Recent枠表示機能(β)とデータ提供機能の組み込み

MAX-eipi opened the pull request:
#36 の対応です。
Recent枠表示機能(β)とデータ提供機能を実装しました。

【Recent枠表示機能(β)について】
・クライアント内で閉じた機能です。
・各種通信APIを呼び出すところでプレイ履歴取得APIを呼び出す必要があります。
・Recent枠が単曲レート値降順で表示され、その下にRecent候補枠がプレイ日時降順で表示されます。

【データ提供機能について】
・サーバーと通信する機能です。
・実行するとユーザー情報やレコード、プレイ履歴等の情報を取得して、GASにデータを送ります。
・サーバー側でシミュレートしたRecent候補枠とRecent枠の結果が渡され、アラートで表示されます。
・クライアント側の処理はフォークした私のリポジトリ内に配置されています。

ご確認の程、よろしくお願いします。

hh-pg-math committed:
Recent枠表示機能(β)のシステム実装

hh-pg-math committed:
Recent枠表示機能(β)の表示処理実装

hh-pg-math committed:
データ提供機能の組み込み

hh-pg-math committed:
更新文言の追加

zk-phi commented:
ありがとうございます><
早めにレビューします！

zk-phi commented:
`include` している先のコードまでちゃんと見たいので、ちょっとかかるかもですmm

zk-phi commented on chunithm.js:
```diff
@@ -1129,6 +1130,32 @@ function rate_display()
         });
 
+        $("#cra_manage_play_data").click(function () {
+            function includes(root, files) {
```
`dependencies` でも同じような処理を書いているので、できれば使いまわしたい感があります
（順番にロードしないとまずいからこうなってる感じですかね…？）

MAX-eipi replied:
そうです。プリロードが完了しないと動作しないため、そのように実装しております。
しかしそれは、ブックマークレット内の処理をそのまま持ってきたためであり、
付けくわえるとdependenciesの存在を見落としておりました、、。

dependenciesで読み込むことを前提とした場合、次の様に変更することが可能です。
・dependenciesに gas_data_manager.js と manage_histroy.js を追加する。
・ボタンクリック時の処理を exec_manage_history.js 内の処理に置き換える。

この対応で問題なければ、そのように修正することを試みますが、いかがでしょうか？
ご検討の程、よろしくお願いします。

zk-phi replied:
それぞれのファイルまだちゃんと読めてないのでそれから考えてみますねー。

zk-phi replied:
gaslibs の方はまた改めてちゃんと見ますが、ざっくり見た感じそこまでの規模でもなさそうなので、１ファイルにまとめてしまって

```
$('head').append("<script src='https://max-eipi.gihub.io/CHUNITHMRateAnalyzer/data_submit.js'>")
```

みたいな感じで一撃で読んでしまうのがよさげですかねー。

zk-phi commented on chunithm.js:
```diff
@@ -416,6 +416,8 @@ var opt_rate    = 0;
 var disp_rate   = 0;
 var recent_rate = 0;
 var worst_chart_rate;
+var recent_candidate_list = new Array();
```
`recent_candidates` とかちょっと短めの名前にして `=` の位置を揃えたい感じがあります (`last_recent_candidate_list` も同様ですね)

MAX-eipi replied:
ご提案通り、
recent_candidate_list -> recent_candidates
last_recent_candidate_list -> last_recent_candidates
に変更しようと思います。

zk-phi replied:
:+1:

zk-phi commented on chunithm.js:
```diff
@@ -632,6 +642,82 @@ $("#cra_wrapper").delay(400).fadeIn(400);
+{
+    $("#cra_window_inner").html("<p>loading playlog ...</p>");
+    request_api("GetUserPlaylogApi", {}, function (d) {
+        var level_name_map = ["basic", "advance", "expert", "master", "worldsend"];
```
もし `d.userPlaylogList[i].level` が提供されているならいらなそうです (`playlogapi` 直接叩いたことないので `levelName` しか提供されていないようならスルーで ok です)

MAX-eipi replied:
userPlaylogList には level が含まれていないので、ここは現状維持になるかと思われます。

zk-phi replied:
了解ですー

zk-phi commented on chunithm.js:
```diff
@@ -632,6 +642,82 @@ $("#cra_wrapper").delay(400).fadeIn(400);
+            if (d.userPlaylogList[i].levelName != "expert" && d.userPlaylogList[i].levelName != "master")
+                continue;
+
+            var date_info = /(\d{4})-(\d{2})-(\d{2}) (\d{1,2}):(\d{2}):(\d{2}).\d*/.exec(d.userPlaylogList[i].userPlayDate);
```
大小比較したいだけなら Date に変換しなくても、単純な文字列比較でいけるかなと思いました

MAX-eipi replied:
jsでは文字列を辞書順で大小比較可能でしたでしょうか?
そうであれば、仰る通り文字列比較に切り替えます。

zk-phi replied:
<で辞書順比較になったと思います(うろ覚えなので念のため確認お願いしますmm)

zk-phi replied:
:+1:

zk-phi commented on chunithm.js:
```diff
@@ -632,6 +642,82 @@ $("#cra_wrapper").delay(400).fadeIn(400);
+
+        recent_candidate_list = last_recent_candidate_list;
+        if (!recent_candidate_list) {
+            recent_candidate_list = new Array();
```
```
Array.apply(null,  new(array(30))).map(function(x) { return { rate_base: 0,  ... } })
```
という裏技がありますが、どっちが読みやすいかはきわどいですね…。配列が初期化されていなくても大丈夫なようにほかのコードを書いて、ここでは初期化しないというのも手ですね。一番読みやすいと思う方法でいいと思います

MAX-eipi replied:
ここはなるべく簡潔にしたいです。
私としてはご提案いただいた方法でも十分に感じましたが、少し考察してみます。

zk-phi replied:
:+1:
これは余談ですが、 `function (_) { return create_playlog(0, 0, ""); }` は `create_playlog.bind(null, 0, 0, "")` と書くこともできます (まあどっちもどっちですね…ｗ こっちのほうが関数型っぽくて好みだったりします)

zk-phi commented on chunithm.js:
```diff
@@ -632,6 +642,82 @@ $("#cra_wrapper").delay(400).fadeIn(400);
+            if (play_date <= last_play_date)
+                continue;
+
+            for (var j = 0; j < chart_list.length; j++) {
```
この二重ループはちょっと計算コストが重いですね…（仕方ないですが；；）
欲を言えば、 chart_list を曲名をキーにしたハッシュにして、一撃で探せるようにしたいとこですね…

zk-phi replied:
(僕の設計ミスです)

zk-phi replied:
あ、赤譜面だと `chart_list[i].name` に `" 赤"` がくっついてしまっているのでそれもまずいですね…
`chart_list` の持ち方はちょっと考えないといけなそうです

zk-phi replied:
例えばですが、recent 枠だろうが best枠だろうが曲名、ジャケ画像、譜面定数とかの情報は変わらないので、

```
// 曲の情報
var chart_list = {
    ....
    10: { name: 'いい曲', image: 'path/to/img.png', rate_base: { 2: 11.0, 3: 13.0 } }, 
     ...
}

// ベスト枠
best_rate_list = [ { chart_id: 10, score: 10000000, rate: 14.0 }, ... ];

// Recent 枠＆ Recent 候補枠
recent_rate_list = [ { chart_id: 234, score: 9000000, rate: 13.2 }, ... ];
```
みたいな感じで曲の情報とスコアを別別に持つのも手かもしれないですね…（ちょっと影響範囲大きいですが）

MAX-eipi replied:
私も少し前からそういうデータ構造にした方が、
データ検索効率や、今後の楽曲追加の作業効率の向上が見込めるのではないかと考えておりました。
頃合いをみてデータ構造を変えてみても良いかと思います。
ただ、大きく改修する必要があるため、実装との整合性を配慮し、
その作業自体はそちらにお任せしたいところではあります。

zk-phi commented on chunithm.js:
```diff
@@ -632,6 +642,82 @@ $("#cra_wrapper").delay(400).fadeIn(400);
+                        play_date: play_date
+                    };
+
+                    recent_list = [].concat(recent_candidate_list).sort(function (p1, p2) {
```
１曲追加するごとに map や sort するのもちょっとオーバーキル気味な感じがあります。
たとえばレート `a` の曲がリストから抜けてレート `b` の曲がリストに入った場合、新しいレートは元のレートに `(b - a) / 30` を足したものになりそうですから、リスト全体を舐めなくても計算できそうです。

MAX-eipi replied:
確かにオーダーがかさばるので、対処すべき点ではあるのですが、
実装当時は他に良い方法が思いつきませんでした。
ここでRecent候補枠からRecent枠を抽出しているのは、
最小単曲レート値と最小スコアを取得するためですので、
その2値をより効率的に取得できる方法が思いつけばという所存です。

zk-phi replied:
差分計算で頑張ったらどうなるかちょっと考えてみたのですが、ひと手間かけないと(配列ではダメ)たかだかlogN倍しか速くできなさそうだったので、そこまで頑張らなくても良さそうです。また暇なときに考えます

zk-phi commented on chunithm.js:
```diff
@@ -605,10 +609,14 @@ $("#cra_window_inner")
-                            localStorage.setItem("cra_version", JSON.stringify(cra_version));
-                            $("#cra_close_button").show(400);
-                            rate_display();
+                            fecth_playlog(function () {
```
`fetch` の typo です

zk-phi replied:
:+1:

zk-phi commented on chunithm.js:
```diff
@@ -632,6 +642,82 @@ $("#cra_wrapper").delay(400).fadeIn(400);
+                            }
+                        }
+                    }
+                    else if (playlog.score >= 1007500 || playlog.score >= Math.min.apply(null, recent_list.map(function (p) { return p.score; }))) {
```
```
else if (hoge) { }
else {
    piyo;
}
```
は
```
else if (!hoge) {
    piyo
}
```
にできそうです

zk-phi replied:
:+1:

zk-phi commented:
71ab193 まで読みました :+1:

zk-phi commented:
71ab193 まで読みました 👍

zk-phi commented on chunithm.js:
```diff
@@ -1338,3 +1347,120 @@ function render_chart_list(msgs)
 }
+
+// refresh the recent list display
+function render_recent_list(msgs1, msgs2)
```
コードの大部分が `render_chart_list` と同じなのでうまく共通化したいところですね。
recent 枠と recent 枠を１本の長い配列として持っておいて (そのうち先頭から10曲がRecent枠)、
```
// Best 枠をレンダリングするとき
render_chart_list(chart_list,  { 0: 'Best 枠',  30: 'Best 枠ここまで' })
// Recent 枠をレンダリングするとき
render_chart_list(recent_candidates, { 0: 'Recent 枠', 10: 'Recent候補枠' })
```
のように配列を引数に取るようにするとかどうでしょう？

MAX-eipi replied:
共通化することに関して同感です。
ご提案通りにできればと存じます。

zk-phi replied:
これは chart_list まわりの改修のついでにやっちゃいますね

zk-phi commented on chunithm.js:
```diff
@@ -587,6 +587,7 @@ $("#cra_window_outer")
     .html("<div id='cra_window_inner' class='frame01_inside w450'></div>");
 $("#cra_window_inner")
     .html("<p class='cra_caution'>CAUTION</p>" +
+          "<p>12/3- Recent枠表示機能(β)とデータ提供機能の追加。</p>" +
```
せっかくなので `提供：<a href="twitter.com/max_eipi">@max_eipi</a>` とか付けてもらっちゃってもいいですよ！お任せします

zk-phi commented:
`gaslibs` 以外読みました :+1:

zk-phi commented on chunithm.js:
```diff
@@ -1062,7 +1062,8 @@ function rate_display()
                   "<div id='cra_sort_score_req' class='cra_sort_button'>必要スコア順</div>" +
-                  "<div id='cra_sort_score_ave' class='cra_sort_button'>おすすめ(β)</div>");
+                  "<div id='cra_sort_score_ave' class='cra_sort_button'>おすすめ(β)</div>" +
+                  "<div id='cra_manage_play_data' class='cra_sort_button'>プレイデータ管理</div>");
```
基本的には eipi さんにデータを提供するための機能になると思うので、「プレイデータ管理」という名前はちょっと紛らわしいかもしれないですね（押してしまうとがっつり通信が発生してしまうので、不親切かもです）。
たとえばですが、タイトルのメッセージ画面に「Recent 枠の調査にご協力ください！」的な感じで説明を書いておいて、スコア一覧画面の欄外に「Google スプレッドシートにエクスポート」みたいな感じでリンクを出しておくとか、そんな感じで設置するのが（ただレートの内訳を知りたいだけの人にとっては）親切かもしれないですかね…。

MAX-eipi replied:
そうですね。
ここに何も考えずにボタン配置したのは正直愚策だとは考えています。
なので、希望としてはソートボタンから分離した上で、上記のような対応ができたらと思います。

また、最近通信にかなり時間がかかっているので、レスポンス無視でクライアント側からデータが完了したら「送信完了」などのメッセージを表示してアナライザーに復帰するようにした方が待機時間も削れるかと思いました。(レスポンス無視する以上Recent枠がアナライザー上で確認できなくなるのですが、、)

MAX-eipi replied:
クライアント側からデータの送信が完了したら

ですね。

MAX-eipi commented:
レビューありがとうございます。
そちらのリファクタリング作業との兼ね合いがありますと思いますが、それが終わるまで待った方がいいですか？

zk-phi commented:
あっちは何とかするので気にせず！

MAX-eipi commented:
承知です。
一先ず各個レビューに対応していきます。

hh-pg-math committed:
Merge remote-tracking branch 'host/gh-pages' into plugins

hh-pg-math committed:
変数名の変更
recent_candidate_list -> recent_candidates
last_recent_candidate_list -> last_recent_candidates

hh-pg-math committed:
プレイ時間比較を文字列比較に変更

hh-pg-math committed:
プレイ履歴データ作成処理の修正

hh-pg-math committed:
タイポ修正

hh-pg-math committed:
遷移法則実装のリファクタリング

hh-pg-math committed:
配列初期化処理の修正

hh-pg-math committed:
Recent枠抽出メソッドの修正

zk-phi commented:
差分見ますー。
では、chart_list周りの修正はこちらでやってしまいますね！

hh-pg-math committed:
Recent枠抽出メソッドの修正

Commit:
chart_list 周りの Refactoring

Commit:
Merge branch 'gh-pages' into eipi/plugins

Commit:
Fix criteria for recent-list

Commit:
Fix syntactic errors

Commit:
Fix symbol definition shadowing

Commit:
Fix typos

Commit:
Fix \'undefined\' errors

Commit:
Fix not loading last recent_candidates bug

Commit:
Upcase constants

Commit:
Fetch song titles via API

Commit:
Refactoring

Commit:
Fix recent candidate criteria

Commit:
Fix renderer bug

Commit:
Bump version

Commit:
Merge branch 'gh-pages' of https://github.com/zk-phi/CHUNITHMRateAnalyzer into eipi/plugins

Commit:
Fix base issue

Commit:
Fix recent candidate criteria

Commit:
Remove unused branch

Commit:
Fix rate_diff computation

Commit:
Fix render_chart_list skip chart bug

Commit:
Fix possible reference-to-nulls

hh-pg-math committed:
Merge pull request #1 from zk-phi/eipi/plugins

chart_list 周りの Refactoring

MAX-eipi commented:
TODO
- 譜面定数情報のない楽曲のRecent枠遷移処理の考察
- 現在配置している[プレイデータ管理]の扱い方の再考

hh-pg-math committed:
プレイデータ提供機能の呼び出し処理最適化

hh-pg-math committed:
譜面定数情報が存在しない楽曲のチェック処理追加

MAX-eipi commented:
譜面定数情報のない楽曲は、譜面定数と単曲レート値を0に設定したlogオブジェクトを生成するようにしました。
(コードの改変が少なく済むようにすることを意識したためこのような方針を取りました)

また、[プレイデータ管理]ボタンの表記を[プレイデータ提供]に変更しました。
この、[プレイデータ提供]ボタンをどのように扱うか、いくつか自分の中で案があるため、意見を仰ぎたいと存じます。
[案1] ボタンを押したときにダイアログを表示する(添付画像1枚目)
[案2] ボタンの位置を移動させる
(1) ツイートボタンの横に配置(添付画像2枚目)
(2) フォローボタンの下に配置(添付画像3枚目)
[案3] 案1と案2の合成
イメージとしていくつかレイアウト画像を添付します。参考として御覧ください。

よろしくお願いします。

![1](https://cloud.githubusercontent.com/assets/9124483/21759810/76b65c32-d68a-11e6-9ff7-3c42b0f5683b.png)
![2-1](https://cloud.githubusercontent.com/assets/9124483/21759812/76b963d2-d68a-11e6-8d6e-796c5f4994f7.png)
![2-2](https://cloud.githubusercontent.com/assets/9124483/21759811/76b74a20-d68a-11e6-9136-5b1f1087c053.png)

zk-phi commented:
ダミーデータ作るところで `playlog(...)` の抽象化が効いていないのがちょっと気になりますが、いい代案も思いつかないですし、またリファクタリングの時にきれいにするということで、とりあえずこれでいってしまうのもいいと思います。

プレーデータ提供のコードが chunithm.js と offer_playlog.js の両方に分散してしまっているので、たとえば `gas_manage_history(...)` の大きな式をまるっと
```
function offer_playlog () {
    gas_manage_history(...);
}
```
のような関数として offer_playlog.js 側に移して、メンテが必要になったときに offer_playlog.js だけいじればokな状態にしておくのが便利かなと思いました。

プレーデータ提供は案１と２の折衷に＋１票です。並べ替えメニューに置いてしまうとただ自分のレートを確認したいだけの人にとってはクドく見えてしまうかもしれないですが、ページの一番下だと気づかれない場合も多そうでもったいないかなと思います。説明はいづれにしても必要そうです。

zk-phi commented:
あ、あと「開発者にお知らせください」の箇所は、僕に質問がきてもデータ提供機能に関しては答えられないかもしれないので、 max-eipi さんのほうに届くように誘導してもらったほうがスムーズかなと思いました。
他いいと思います :+1:

zk-phi committed:
Merge branch 'gh-pages' into plugins

hh-pg-math committed:
データ提供機能のリファクタリング

zk-phi commented:
UIまわりクソコードもいいとこなので、説明のダイアログは大変そうなら僕が実装しておきますー。

MAX-eipi commented:
一見した限りでは実装できなくもなさそうですが、他との要因を把握しきれていないため、
実装をお任せしたいと思います。

zk-phi commented:
了解です。スコア提供まわりのコードをoffer_playlogに移すところだけよろしくお願いします、UIの方はこちらでやってしまいますねー。

MAX-eipi commented:
offer_playlog() としてラップして、offer_playlog.js の方に移しておきました。

zk-phi commented:
ありがとうございます。UI周りと #59 の取り込みやっておきますねー

Commit:
Merge branch 'gh-pages' of https://github.com/zk-phi/CHUNITHMRateAnalyzer into plugins

Commit:
データ提供ボタンを移動

zk-phi commented:
ボタンの移動・新曲の取り込み done です。
UI 周りはいづれリファクタリングしたいと思っているのでダイアログは仮にブラウザ標準のダイアログを使いました。
確認お願いしますー、問題なさそうであればリリースしてしまいます。

Commit:
注意書きを追加

MAX-eipi commented:
実装変更箇所の確認と動作確認を行いました。
現状目立った不具合は特に発生しておりません。
自分としては、これで一段落つけてもよいと思います。
よろしくお願いします。

zk-phi commented:
では出してしまいます:+1:
念のためちまちまエゴサしてやばそうだったら引っ込めますね。
